onmt/trainer.py

```diff
@@ -133,79 +152,57 @@ def train(self,
-                                logger.info('GpuRank %d: gather valid stat \
-                                            step %d' % (self.gpu_rank, step))
-                            valid_stats = self._maybe_gather_stats(valid_stats)
-                            if self.gpu_verbose_level > 0:
-                                logger.info('GpuRank %d: report stat step %d'
-                                            % (self.gpu_rank, step))
-                            self._report_step(self.optim.learning_rate(),
-                                              step, valid_stats=valid_stats)
-
-                        if self.gpu_rank == 0:
-                            self._maybe_save(step)
-                        step += 1
-                        if step > train_steps:
-                            break
+        if self.n_gpu > 1:
+            train_iter = itertools.islice(
+                train_iter, self.gpu_rank, None, self.n_gpu)
+
+        for i, (batches, normalization) in enumerate(
+                self._accum_batches(train_iter)):
+            if self.gpu_verbose_level > 1:
+                logger.info("GpuRank %d: index: %d", self.gpu_rank, i)
             if self.gpu_verbose_level > 0:
-                logger.info('GpuRank %d: we completed an epoch \
-                            at step %d' % (self.gpu_rank, step))
+                logger.info("GpuRank %d: reduce_counter: %d \
+                            n_minibatch %d"
+                            % (self.gpu_rank, i + 1, len(batches)))
+
+            if self.n_gpu > 1:
+                normalization = sum(onmt.utils.distributed
+                                    .all_gather_list
+                                    (normalization))
+
+            self._gradient_accumulation(
+                batches, normalization, total_stats,
+                report_stats)
+
+            report_stats = self._maybe_report_training(
+                step, train_steps,
+                self.optim.learning_rate(),
+                report_stats)
+
+            if valid_iter is not None and step % valid_steps == 0:
+                if self.gpu_verbose_level > 0:
+                    logger.info('GpuRank %d: validate step %d'
+                                % (self.gpu_rank, step))
+                valid_stats = self.validate(valid_iter)
+                if self.gpu_verbose_level > 0:
+                    logger.info('GpuRank %d: gather valid stat \
+                                step %d' % (self.gpu_rank, step))
+                valid_stats = self._maybe_gather_stats(valid_stats)
+                if self.gpu_verbose_level > 0:
+                    logger.info('GpuRank %d: report stat step %d'
+                                % (self.gpu_rank, step))
+                self._report_step(self.optim.learning_rate(),
+                                  step, valid_stats=valid_stats)
+
+            if self.gpu_rank == 0:
+                self._maybe_save(step)
+            step += 1
```

Comment:
Is it necessary to keep the `step` variable? I believe `self.optim.training_step` will increment itself correctly. I was going to ask about the break statement too, but I think it is correct now. If I recall, previously it was unnecessary and possibly unreachable because of the while loop.

Reply:
We could certainly do without but it is used several times in the loop so a variable is still useful. Also, I did not want to modify too much the loop content in this PR. Maybe for another one.

Reply:
If is defined inside the loop instead of outside of it, then it doesn't need to be incremented on every iteration.

Reply:
Ok, let's do that as it appears simple/safe enough.